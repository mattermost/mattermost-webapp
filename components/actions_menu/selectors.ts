// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {createSelector} from 'reselect';

import {AppBinding} from '@mattermost/types/apps';

import {AppBindingLocations} from 'mattermost-redux/constants/apps';
import {makeAppBindingsSelector, makeRHSAppBindingSelector} from 'mattermost-redux/selectors/entities/apps';

import {GlobalState} from 'types/store';
import {Locations} from 'utils/constants';

export function makeGetPostOptionBinding(): (state: GlobalState, location?: string) => AppBinding[] | null {
    const centerBindingsSelector = makeAppBindingsSelector(AppBindingLocations.POST_MENU_ITEM);
    const rhsBindingsSelector = makeRHSAppBindingSelector(AppBindingLocations.POST_MENU_ITEM);
    return createSelector(
        'postOptionsBindings',
        centerBindingsSelector,
        rhsBindingsSelector,
        (state: GlobalState, location?: string) => location,
        (centerBindings: AppBinding[], rhsBindings: AppBinding[], location?: string) => {
            switch (location) {
            case Locations.RHS_ROOT:
            case Locations.RHS_COMMENT:
                return rhsBindings;
            case Locations.SEARCH:
                return null;
            case Locations.CENTER:
            default:
                return centerBindings;
            }
        },
    );
}
