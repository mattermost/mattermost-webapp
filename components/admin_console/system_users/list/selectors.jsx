// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getUser, getProfiles, getProfilesInTeam, getProfilesWithoutTeam, searchProfiles, searchProfilesInTeam} from 'mattermost-redux/selectors/entities/users';

const ALL_USERS = '';
const NO_TEAM = 'no_team';
const USER_ID_LENGTH = 26;

export function getUsers(state, loading, teamId, term) {
    if (loading) {
        // Show no users while loading.
        return [];
    }

    if (term) {
        let users = [];
        if (teamId) {
            users = searchProfilesInTeam(state, teamId, term);
        } else {
            users = searchProfiles(state, term);
        }

        if (users.length === 0 && term.length === USER_ID_LENGTH) {
            const user = getUser(state, term);
            if (user) {
                users = [user];
            }
        }

        return users;
    }

    if (teamId === ALL_USERS) {
        return getProfiles(state);
    } else if (teamId === NO_TEAM) {
        return getProfilesWithoutTeam(state);
    }

    return getProfilesInTeam(state, teamId);
}
