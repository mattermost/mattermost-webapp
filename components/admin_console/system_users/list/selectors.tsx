// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getUser, getProfiles, getProfilesInTeam, getProfilesWithoutTeam, makeSearchProfilesStartingWithTerm, searchProfilesInTeam} from 'mattermost-redux/selectors/entities/users';
import {GlobalState} from 'mattermost-redux/types/store';
import {UserProfile} from 'mattermost-redux/types/users';

import {userSelectorOptionsFromFilter} from 'utils/filter_users';

const ALL_USERS = '';
const NO_TEAM = 'no_team';
const USER_ID_LENGTH = 26;

export function getUsers(state: GlobalState, loading: boolean, teamId: string, term: string, filter: string): UserProfile[] {
    if (loading) {
        // Show no users while loading.
        return [];
    }
    const searchProfilesStartingWithTerm = makeSearchProfilesStartingWithTerm();

    const filters = userSelectorOptionsFromFilter(filter);
    if (term) {
        let users = [];
        if (teamId) {
            users = searchProfilesInTeam(state, teamId, term, false, filters);
        } else {
            users = searchProfilesStartingWithTerm(state, term, false, filters);
        }

        if (users.length === 0 && term.length === USER_ID_LENGTH) {
            const user = getUser(state, term);
            if (user) {
                users = [user];
            }
        }

        return users;
    }

    if (teamId === ALL_USERS) {
        return getProfiles(state, filters);
    } else if (teamId === NO_TEAM) {
        return getProfilesWithoutTeam(state, filters);
    }

    return getProfilesInTeam(state, teamId, filters);
}

export function getNonBotUsers(state: GlobalState, loading: boolean, teamId: string, term: string, filter: string): UserProfile[] {
    return getUsers(state, loading, teamId, term, filter).filter((user) => {
        return !user.is_bot;
    });
}
