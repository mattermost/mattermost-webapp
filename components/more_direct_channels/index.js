// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {getProfiles, getProfilesInTeam, searchProfiles} from 'mattermost-redux/actions/users';
import {
    getCurrentUserId,
    getProfiles as selectProfiles,
    getProfilesInCurrentChannel,
    getProfilesInCurrentTeam,
    searchProfiles as searchProfilesSelector,
    searchProfilesInCurrentTeam,
} from 'mattermost-redux/selectors/entities/users';
import {getConfig} from 'mattermost-redux/selectors/entities/general';
import {getCurrentTeam} from 'mattermost-redux/selectors/entities/teams';

import {setModalSearchTerm} from 'actions/views/search';

import MoreDirectChannels from './more_direct_channels.jsx';

function mapStateToProps(state, ownProps) {
    let currentChannelMembers = [];
    if (ownProps.isExistingChannel) {
        currentChannelMembers = getProfilesInCurrentChannel(state);
    }

    const config = getConfig(state);
    const restrictDirectMessage = config.RestrictDirectMessage;

    const searchTerm = state.views.search.modalSearch;

    let users;
    if (searchTerm) {
        if (restrictDirectMessage === 'any') {
            users = searchProfilesSelector(state, searchTerm, false);
        } else {
            users = searchProfilesInCurrentTeam(state, searchTerm, false);
        }
    } else if (restrictDirectMessage === 'any') {
        users = selectProfiles(state);
    } else {
        users = getProfilesInCurrentTeam(state);
    }

    const team = getCurrentTeam(state);

    return {
        currentTeamId: team.id,
        currentTeamName: team.name,
        searchTerm,
        users,
        statuses: state.entities.users.statuses,
        currentChannelMembers,
        currentUserId: getCurrentUserId(state),
        restrictDirectMessage,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            getProfiles,
            getProfilesInTeam,
            searchProfiles,
            setModalSearchTerm,
        }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(MoreDirectChannels);
