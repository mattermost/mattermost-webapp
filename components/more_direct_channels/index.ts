// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';
import {bindActionCreators, ActionCreatorsMapObject, Dispatch} from 'redux';
import {
    getProfiles,
    getProfilesInTeam,
    getStatusesByIds,
    getTotalUsersStats,
    searchProfiles,
} from 'mattermost-redux/actions/users';
import {searchGroupChannels} from 'mattermost-redux/actions/channels';
import {
    getCurrentUserId,
    getProfiles as selectProfiles,
    getProfilesInCurrentChannel,
    getProfilesInCurrentTeam,
    makeSearchProfilesStartingWithTerm,
    searchProfilesInCurrentTeam,
    getTotalUsersStats as getTotalUsersStatsSelector,
} from 'mattermost-redux/selectors/entities/users';
import {getChannelsWithUserProfiles, getAllChannels} from 'mattermost-redux/selectors/entities/channels';
import {getConfig} from 'mattermost-redux/selectors/entities/general';
import {getCurrentTeam} from 'mattermost-redux/selectors/entities/teams';
import {ActionFunc, GenericAction} from 'mattermost-redux/types/actions';
import {Channel} from 'mattermost-redux/types/channels';
import {UserProfile} from 'mattermost-redux/types/users';
import {sortByUsername, filterProfilesStartingWithTerm} from 'mattermost-redux/utils/user_utils';
import {memoizeResult} from 'mattermost-redux/utils/helpers';

import {openDirectChannelToUserId, openGroupChannelToUserIds} from 'actions/channel_actions';
import {loadStatusesForProfilesList} from 'actions/status_actions.jsx';
import {loadProfilesForGroupChannels} from 'actions/user_actions.jsx';
import {setModalSearchTerm} from 'actions/views/search';

import {GlobalState} from 'types/store';

import MoreDirectChannels from './more_direct_channels';

type OwnProps = {
    isExistingChannel: boolean;
}

const makeMapStateToProps = () => {
    const searchProfilesStartingWithTerm = makeSearchProfilesStartingWithTerm();

    return (state: GlobalState, ownProps: OwnProps) => {
        const currentUserId = getCurrentUserId(state);
        let currentChannelMembers: UserProfile[] = [];
        if (ownProps.isExistingChannel) {
            currentChannelMembers = getProfilesInCurrentChannel(state);
        }

        const config = getConfig(state);
        const restrictDirectMessage = config.RestrictDirectMessage;

        const searchTerm = state.views.search.modalSearch;

        let users;
        if (searchTerm) {
            if (restrictDirectMessage === 'any') {
                users = searchProfilesStartingWithTerm(state, searchTerm, false);
            } else {
                users = searchProfilesInCurrentTeam(state, searchTerm, false);
            }
        } else if (restrictDirectMessage === 'any') {
            users = selectProfiles(state, {});
        } else {
            users = getProfilesInCurrentTeam(state);
        }

        const filteredGroupChannels = filterGroupChannels(getChannelsWithUserProfiles(state), searchTerm);
        const myDirectChannels = filterDirectChannels(getAllChannels(state), currentUserId);

        const team = getCurrentTeam(state);
        const stats = getTotalUsersStatsSelector(state) || {total_users_count: 0};

        return {
            currentTeamId: team.id,
            currentTeamName: team.name,
            searchTerm,
            users: users.sort(sortByUsername),
            myDirectChannels,
            groupChannels: filteredGroupChannels,
            statuses: state.entities.users.statuses,
            currentChannelMembers,
            currentUserId,
            restrictDirectMessage,
            totalCount: stats.total_users_count,
        };
    };
};

const filterGroupChannels = memoizeResult((channels: Array<{profiles: Array<UserProfile>} & Channel>, term: string) => {
    return channels.filter((channel) => {
        const matches = filterProfilesStartingWithTerm(channel.profiles, term);
        return matches.length > 0;
    });
});

const filterDirectChannels = memoizeResult((channels: Channel[], userId: string) => {
    return Object.values(channels).filter((channel) => {
        if (channel.type !== 'D') {
            return false;
        }
        if (channel.name && channel.name.indexOf(userId) < 0) {
            return false;
        }
        return true;
    });
});

type Actions = {
    getProfiles: (page?: number | undefined, perPage?: number | undefined, options?: any) => Promise<any>;
    getProfilesInTeam: (teamId: string, page: number, perPage?: number | undefined, sort?: string | undefined, options?: any) => Promise<any>;
    getStatusesByIds: (userIds: string[]) => ActionFunc;
    getTotalUsersStats: () => ActionFunc;
    loadStatusesForProfilesList: (users: any) => {
        data: boolean;
    };
    loadProfilesForGroupChannels: (groupChannels: any) => Promise<any>;
    openDirectChannelToUserId: (userId: any) => Promise<any>;
    openGroupChannelToUserIds: (userIds: any) => Promise<any>;
    searchProfiles: (term: string, options?: any) => Promise<any>;
    searchGroupChannels: (term: string) => Promise<any>;
    setModalSearchTerm: (term: any) => GenericAction;
}

function mapDispatchToProps(dispatch: Dispatch) {
    return {
        actions: bindActionCreators<ActionCreatorsMapObject<ActionFunc | GenericAction>, Actions>({
            getProfiles,
            getProfilesInTeam,
            getStatusesByIds,
            getTotalUsersStats,
            loadStatusesForProfilesList,
            loadProfilesForGroupChannels,
            openDirectChannelToUserId,
            openGroupChannelToUserIds,
            searchProfiles,
            searchGroupChannels,
            setModalSearchTerm,
        }, dispatch),
    };
}

export default connect(makeMapStateToProps, mapDispatchToProps)(MoreDirectChannels);
