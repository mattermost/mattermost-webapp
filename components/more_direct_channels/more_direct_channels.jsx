// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import PropTypes from 'prop-types';
import React from 'react';
import {Modal} from 'react-bootstrap';
import {FormattedMessage} from 'react-intl';
import {Client4} from 'mattermost-redux/client';

import {browserHistory} from 'utils/browser_history';
import Constants from 'utils/constants.jsx';
import {displayEntireNameForUser, localizeMessage, isGuest} from 'utils/utils.jsx';
import MultiSelect from 'components/multiselect/multiselect.jsx';
import ProfilePicture from 'components/profile_picture';
import AddIcon from 'components/widgets/icons/fa_add_icon';
import GuestBadge from 'components/widgets/badges/guest_badge';
import BotBadge from 'components/widgets/badges/bot_badge';

import GroupMessageOption from './group_message_option';

const USERS_PER_PAGE = 50;
const MAX_SELECTABLE_VALUES = Constants.MAX_USERS_IN_GM - 1;

export default class MoreDirectChannels extends React.Component {
    static propTypes = {

        currentUserId: PropTypes.string.isRequired,
        currentTeamId: PropTypes.string.isRequired,
        currentTeamName: PropTypes.string.isRequired,
        searchTerm: PropTypes.string.isRequired,
        users: PropTypes.arrayOf(PropTypes.object).isRequired,
        groupChannels: PropTypes.arrayOf(PropTypes.object).isRequired,
        statuses: PropTypes.object.isRequired,
        totalCount: PropTypes.number,

        /*
         * List of current channel members of existing channel
         */
        currentChannelMembers: PropTypes.arrayOf(PropTypes.object),

        /*
         * Whether the modal is for existing channel or not
         */
        isExistingChannel: PropTypes.bool.isRequired,

        /*
         * The mode by which direct messages are restricted, if at all.
         */
        restrictDirectMessage: PropTypes.string,
        onModalDismissed: PropTypes.func,
        onHide: PropTypes.func,
        bodyOnly: PropTypes.bool,
        actions: PropTypes.shape({
            getProfiles: PropTypes.func.isRequired,
            getProfilesInTeam: PropTypes.func.isRequired,
            getStatusesByIds: PropTypes.func.isRequired,
            getTotalUsersStats: PropTypes.func.isRequired,
            loadStatusesForProfilesList: PropTypes.func.isRequired,
            loadProfilesForGroupChannels: PropTypes.func.isRequired,
            openDirectChannelToUserId: PropTypes.func.isRequired,
            openGroupChannelToUserIds: PropTypes.func.isRequired,
            searchProfiles: PropTypes.func.isRequired,
            searchGroupChannels: PropTypes.func.isRequired,
            setModalSearchTerm: PropTypes.func.isRequired,
        }).isRequired,
    }

    constructor(props) {
        super(props);

        this.searchTimeoutId = 0;

        const values = [];

        if (props.currentChannelMembers) {
            for (let i = 0; i < props.currentChannelMembers.length; i++) {
                const user = Object.assign({}, props.currentChannelMembers[i]);

                if (user.id === props.currentUserId) {
                    continue;
                }

                values.push(user);
            }
        }

        this.state = {
            values,
            show: true,
            search: false,
            saving: false,
            loadingUsers: true,
        };
    }

    componentDidMount() {
        this.getUserProfiles();
        this.props.actions.getTotalUsersStats();
        this.loadProfilesMissingStatus(this.props.users, this.props.statuses);
    }

    UNSAFE_componentWillReceiveProps(nextProps) { // eslint-disable-line camelcase
        if (this.props.searchTerm !== nextProps.searchTerm) {
            clearTimeout(this.searchTimeoutId);

            const searchTerm = nextProps.searchTerm;
            if (searchTerm === '') {
                this.resetPaging();
            } else {
                const teamId = nextProps.restrictDirectMessage === 'any' ? '' : nextProps.currentTeamId;

                this.searchTimeoutId = setTimeout(
                    async () => {
                        this.setUsersLoadingState(true);
                        const [{data: profilesData}, {data: groupChannelsData}] = await Promise.all([
                            this.props.actions.searchProfiles(searchTerm, {team_id: teamId}),
                            this.props.actions.searchGroupChannels(searchTerm),
                        ]);
                        if (profilesData) {
                            this.props.actions.loadStatusesForProfilesList(profilesData);
                        }
                        if (groupChannelsData) {
                            this.props.actions.loadProfilesForGroupChannels(groupChannelsData);
                        }
                        this.resetPaging();
                        this.setUsersLoadingState(false);
                    },
                    Constants.SEARCH_TIMEOUT_MILLISECONDS
                );
            }
        }

        if (
            this.props.users.length !== nextProps.users.length ||
            Object.keys(this.props.statuses).length !== Object.keys(nextProps.statuses).length
        ) {
            this.loadProfilesMissingStatus(nextProps.users, nextProps.statuses);
        }
    }

    loadProfilesMissingStatus = (users = [], statuses = {}) => {
        const missingStatusByIds = users.
            filter((user) => !statuses[user.id]).
            map((user) => user.id);

        if (missingStatusByIds.length > 0) {
            this.props.actions.getStatusesByIds(missingStatusByIds);
        }
    }

    handleHide = () => {
        this.props.actions.setModalSearchTerm('');
        this.setState({show: false});

        if (this.props.bodyOnly) {
            this.handleExit();
        }
    }

    setUsersLoadingState = (loadingState) => {
        this.setState({
            loadingUsers: loadingState,
        });
    }

    handleExit = () => {
        if (this.exitToChannel) {
            browserHistory.push(this.exitToChannel);
        }

        if (this.props.onModalDismissed) {
            this.props.onModalDismissed();
        }

        if (this.props.onHide) {
            this.props.onHide();
        }
    }

    handleSubmit = (values = this.state.values) => {
        const {actions} = this.props;
        if (this.state.saving) {
            return;
        }

        const userIds = values.map((v) => v.id);
        if (userIds.length === 0) {
            return;
        }

        this.setState({saving: true});

        const done = (result) => {
            const {data, error} = result;
            this.setState({saving: false});

            if (!error) {
                this.exitToChannel = '/' + this.props.currentTeamName + '/channels/' + data.name;
                this.handleHide();
            }
        };

        if (userIds.length === 1) {
            actions.openDirectChannelToUserId(userIds[0]).then(done);
        } else {
            actions.openGroupChannelToUserIds(userIds).then(done);
        }
    };

    addValue = (value) => {
        if (Array.isArray(value)) {
            this.addUsers(value);
        } else if ('profiles' in value) {
            this.addUsers(value.profiles);
        } else {
            const values = Object.assign([], this.state.values);

            if (values.indexOf(value) === -1) {
                values.push(value);
            }

            this.setState({values});
        }
    };

    addUsers = (users) => {
        const values = Object.assign([], this.state.values);
        const existingUserIds = values.map((user) => user.id);
        for (const user of users) {
            if (existingUserIds.indexOf(user.id) !== -1) {
                continue;
            }
            values.push(user);
        }

        this.setState({values});
    };

    getUserProfiles = (page) => {
        const pageNum = page ? page + 1 : 0;
        if (this.props.restrictDirectMessage === 'any') {
            this.props.actions.getProfiles(pageNum, USERS_PER_PAGE * 2).then(() => {
                this.setUsersLoadingState(false);
            });
        } else {
            this.props.actions.getProfilesInTeam(this.props.currentTeamId, pageNum, USERS_PER_PAGE * 2).then(() => {
                this.setUsersLoadingState(false);
            });
        }
    }

    handlePageChange = (page, prevPage) => {
        if (page > prevPage) {
            this.setUsersLoadingState(true);
            this.getUserProfiles(page);
        }
    }

    resetPaging = () => {
        if (this.refs.multiselect) {
            this.refs.multiselect.resetPaging();
        }
    }

    search = (term) => {
        this.props.actions.setModalSearchTerm(term);
    }

    handleDelete = (values) => {
        this.setState({values});
    }

    renderAriaLabel = (option) => {
        if (!option) {
            return null;
        }
        return option.username;
    }

    renderOption = (option, isSelected, onAdd) => {
        if (option.type && option.type === 'G') {
            return (
                <GroupMessageOption
                    key={option.id}
                    channel={option}
                    isSelected={isSelected}
                    onAdd={onAdd}
                />
            );
        }

        const displayName = displayEntireNameForUser(option);

        let modalName = displayName;
        if (option.id === this.props.currentUserId) {
            modalName = (
                <FormattedMessage
                    id='more_direct_channels.directchannel.you'
                    defaultMessage='{displayname} (you)'
                    values={{
                        displayname: displayName,
                    }}
                />
            );
        } else if (option.delete_at) {
            modalName = (
                <FormattedMessage
                    id='more_direct_channels.directchannel.deactivated'
                    defaultMessage='{displayname} - Deactivated'
                    values={{
                        displayname: displayName,
                    }}
                />
            );
        }

        var rowSelected = '';
        if (isSelected) {
            rowSelected = 'more-modal__row--selected';
        }

        const status = option.delete_at || option.is_bot ? null : this.props.statuses[option.id];
        const email = option.is_bot ? null : option.email;

        return (
            <div
                key={option.id}
                ref={isSelected ? 'selected' : option.id}
                className={'more-modal__row clickable ' + rowSelected}
                onClick={() => onAdd(option)}
            >
                <ProfilePicture
                    src={Client4.getProfilePictureUrl(option.id, option.last_picture_update)}
                    status={status}
                    size='md'
                />
                <div
                    className='more-modal__details'
                >
                    <div className='more-modal__name'>
                        {modalName}
                        <BotBadge
                            show={Boolean(option.is_bot)}
                            className='badge-popoverlist'
                        />
                        <GuestBadge
                            show={isGuest(option)}
                            className='badge-popoverlist'
                        />
                    </div>
                    <div className='more-modal__description'>
                        {email}
                    </div>
                </div>
                <div className='more-modal__actions'>
                    <div className='more-modal__actions--round'>
                        <AddIcon/>
                    </div>
                </div>
            </div>
        );
    }

    renderValue(props) {
        return props.data.username;
    }

    handleSubmitImmediatelyOn = (value) => {
        return value.id === this.props.currentUserId || Boolean(value.delete_at);
    }

    render() {
        let note;
        if (this.props.currentChannelMembers) {
            if (this.state.values && this.state.values.length >= MAX_SELECTABLE_VALUES) {
                note = (
                    <FormattedMessage
                        id='more_direct_channels.new_convo_note.full'
                        defaultMessage={'You\'ve reached the maximum number of people for this conversation. Consider creating a private channel instead.'}
                    />
                );
            } else if (this.props.isExistingChannel) {
                note = (
                    <FormattedMessage
                        id='more_direct_channels.new_convo_note'
                        defaultMessage={'This will start a new conversation. If you\'re adding a lot of people, consider creating a private channel instead.'}
                    />
                );
            }
        }

        const buttonSubmitText = localizeMessage('multiselect.go', 'Go');
        const buttonSubmitLoadingText = localizeMessage('multiselect.loading', 'Loading..');

        const numRemainingText = (
            <FormattedMessage
                id='multiselect.numPeopleRemaining'
                defaultMessage='Use ↑↓ to browse, ↵ to select. You can add {num, number} more {num, plural, one {person} other {people}}. '
                values={{
                    num: MAX_SELECTABLE_VALUES - this.state.values.length,
                }}
            />
        );

        let users = this.props.users || [];

        if (this.state.values.length) {
            users = users.filter((user) => user.delete_at === 0 && user.id !== this.props.currentUserId);
        } else {
            const active = [];
            const inactive = [];
            for (const user of users) {
                (user.delete_at ? inactive : active).push(user);
            }
            users = active.concat(inactive);
        }

        const groupChannels = this.props.groupChannels || [];

        const options = [...users, ...groupChannels];
        const body = (
            <MultiSelect
                key='moreDirectChannelsList'
                ref='multiselect'
                options={options}
                optionRenderer={this.renderOption}
                values={this.state.values}
                valueRenderer={this.renderValue}
                ariaLabelRenderer={this.renderAriaLabel}
                perPage={USERS_PER_PAGE}
                handlePageChange={this.handlePageChange}
                handleInput={this.search}
                handleDelete={this.handleDelete}
                handleAdd={this.addValue}
                handleSubmit={this.handleSubmit}
                noteText={note}
                maxValues={MAX_SELECTABLE_VALUES}
                numRemainingText={numRemainingText}
                buttonSubmitText={buttonSubmitText}
                buttonSubmitLoadingText={buttonSubmitLoadingText}
                submitImmediatelyOn={this.handleSubmitImmediatelyOn}
                saving={this.state.saving}
                loading={this.state.loadingUsers}
                users={this.props.users}
                totalCount={this.props.totalCount}
                placeholderText={localizeMessage('multiselect.placeholder', 'Search and add members')}
            />
        );

        if (this.props.bodyOnly) {
            return body;
        }

        return (
            <Modal
                dialogClassName='a11y__modal more-modal more-direct-channels'
                show={this.state.show}
                onHide={this.handleHide}
                onExited={this.handleExit}
                role='dialog'
                aria-labelledby='moreDmModalLabel'
                id='moreDmModal'
            >
                <Modal.Header closeButton={true}>
                    <Modal.Title
                        componentClass='h1'
                        id='moreDmModalLabel'
                    >
                        <FormattedMessage
                            id='more_direct_channels.title'
                            defaultMessage='Direct Messages'
                        />
                    </Modal.Title>
                </Modal.Header>
                <Modal.Body
                    role='application'
                >
                    {body}
                </Modal.Body>
                <Modal.Footer className='modal-footer--invisible'>
                    <button
                        id='closeModalButton'
                        type='button'
                        className='btn btn-link'
                    >
                        <FormattedMessage
                            id='general_button.close'
                            defaultMessage='Close'
                        />
                    </button>
                </Modal.Footer>
            </Modal>
        );
    }
}
