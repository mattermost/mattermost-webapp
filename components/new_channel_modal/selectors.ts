// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getConfig} from 'mattermost-redux/selectors/entities/general';
import {isMinimumServerVersion} from 'mattermost-redux/utils/helpers';

import {GlobalState} from 'types/store';

import {suitePluginIds} from 'utils/constants';

export function canCreateBoards(state: GlobalState) {
    const config = getConfig(state);
    if (config.FeatureFlagBoardsProduct === 'true') {
        return true;
    }

    // 7.2.1 is the minimum required; it exposes the boards templates api
    const focalboardManifest = state.plugins.plugins.focalboard;
    return focalboardManifest &&
        focalboardManifest.id === suitePluginIds.focalboard &&
        isMinimumServerVersion(focalboardManifest.version, 7, 2, 1);
}
