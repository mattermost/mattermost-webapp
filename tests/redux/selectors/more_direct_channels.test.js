// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {matchExistsInChannelProfiles, getChannelsWithUserProfiles} from 'selectors/more_direct_channels';

jest.mock('mattermost-redux/selectors/entities/channels', () => ({
    getGroupChannels: jest.fn(() => {
        var channel = {};
        channel.name = 'some-name';
        channel.display_name = `Unit Test ${channel.name}`;
        channel.type = 'G';
        return [channel];
    }),
}));

jest.mock('mattermost-redux/selectors/entities/users', () => ({

    makeGetProfilesInChannel: jest.fn(() => {
        const mockUser1 = {id: 'user1', password: 'password', username: 'user1'};
        const mockUser2 = {id: 'user2', password: 'password', username: 'user2'};
        const mockUser3 = {id: 'user3', password: 'password', username: 'user3'};
        return () => [mockUser1, mockUser2, mockUser3];
    }),
    getCurrentUserId: jest.fn(() => 'user1'),
}));

describe('more_direct_channels selector', () => {
    test('getChannelsWithUserProfiles', () => {
        expect(getChannelsWithUserProfiles({})).toMatchSnapshot();
    });
});

describe('matchExistsInChannelProfiles', () => {
    const mockUser1 = {id: 'user1', password: 'password', username: 'user1', first_name: 'foo', last_name: 'bar', nickname: 'foobar'};
    const mockUser2 = {id: 'user2', password: 'password', username: 'Hello', first_name: 'john', last_name: 'smith', nickname: 'jsmith'};
    const mockUser3 = {id: 'user3', password: 'password', username: 'HelloWorld', first_name: 'steve', last_name: 'fleming', nickname: 'sfleming'};

    const profiles = [mockUser1, mockUser2, mockUser3];

    test('should return true for username', () => {
        const searchTerm = 'Hello';
        const exists = matchExistsInChannelProfiles(profiles, searchTerm);
        expect(exists).toBeTruthy();
    });

    test('should return true for firstName', () => {
        const searchTerm = 'john';
        const exists = matchExistsInChannelProfiles(profiles, searchTerm);
        expect(exists).toBeTruthy();
    });

    test('should return true for lastname', () => {
        const searchTerm = 'fleming';
        const exists = matchExistsInChannelProfiles(profiles, searchTerm);
        expect(exists).toBeTruthy();
    });

    test('should return true for nickname', () => {
        const searchTerm = 'foobar';
        const exists = matchExistsInChannelProfiles(profiles, searchTerm);
        expect(exists).toBeTruthy();
    });

    test('should return false if nothing matches', () => {
        const searchTerm = 'non-existing';
        const exists = matchExistsInChannelProfiles(profiles, searchTerm);
        expect(exists).toBeFalsy();
    });
});

