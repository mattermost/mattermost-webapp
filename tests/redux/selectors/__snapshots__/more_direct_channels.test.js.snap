// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`more_direct_channels selector getChannelsWithUserProfiles 1`] = `
Array [
  Object {
    "display_name": "Unit Test some-name",
    "name": "some-name",
    "profiles": Array [
      Object {
        "id": "user2",
        "password": "password",
        "username": "user2",
      },
      Object {
        "id": "user3",
        "password": "password",
        "username": "user3",
      },
    ],
    "type": "G",
  },
]
`;
